index.js

```diff
@@ -791,39 +791,26 @@ function increment_headers (headers) {
 
 
 
-function step_through_callbacks (callbacks, args, extra, done, index) {
+function step_through_callbacks(callbacks, args, extra, done) {
     args = args || {}
     extra = extra || {}
+    callbacks = (Array.isArray(callbacks)) ? callbacks : [callbacks]
 
-    if (!index) {
-        index = 0
-
-        if (!Array.isArray(callbacks)) {
-            return callbacks(args, done, extra)
-        }
+    let callback = callbacks.shift()
 
-        if (callbacks.length === 1) {
-            return callbacks[0](args, done, extra)
-        }
-
-        return callbacks[index](args, (err, args) => {
-            if (err) {
-                return done(err, args)
-            }
-
-            return step_through_callbacks(callbacks, args, extra, done, ++index)
-        }, extra)
+    if (typeof callback !== 'function') {
+        throw new Error('Callback is not a function')
     }
 
-    if (!callbacks[index]) {
-        return done(null, args)
-    }
-
-    return callbacks[index](args, (err, args) => {
+    return callback(args, (err, args) => {
         if (err) {
             return done(err, args)
         }
 
-        return step_through_callbacks(callbacks, args, extra, done, ++index)
+        if (callbacks.length === 0) {
+            return done(null, args)
```

Comment:
Can we _also_ return {{err}} along with {{args}} here too? It might be that `false` is being passed (or if it's the last function in the chain they could even be passing no error and using {{0}} or something).